src/Microsoft.AspNetCore.JsonPatch/Internal/ParsedPath.cs

```diff
@@ -12,14 +14,19 @@ public struct ParsedPath
 
         private readonly string[] _segments;
 
+        private static readonly Regex escapingRegex = new Regex("~(?<code>0|1)", RegexOptions.Compiled);
+
         public ParsedPath(string path)
         {
             if (path == null)
             {
                 throw new ArgumentNullException(nameof(path));
             }
 
-            _segments = path.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
+            _segments = path
+                .Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries)
+                .Select(segment => escapingRegex.Replace(segment, (match) => match.Groups["code"].Value == "0" ? "~" : "/"))
```

Comment:
Fixed a call to a mysterious Enumerable.Append to List.Add and it's now obvious that the custom implementation is giving better results, So I'm pushing this approach to the branch.
```
RUN 1:
ParsePathWithSplitAndRegex :214 ms.
ParsePathWithSplitAndReplace :77 ms.
ParsePathWithCustomImplementation :27 ms.
RUN 2:
ParsePathWithSplitAndRegex :132 ms.
ParsePathWithSplitAndReplace :74 ms.
ParsePathWithCustomImplementation :31 ms.
RUN 3:
ParsePathWithSplitAndRegex :168 ms.
ParsePathWithSplitAndReplace :79 ms.
ParsePathWithCustomImplementation :26 ms.
RUN 4:
ParsePathWithSplitAndRegex :129 ms.
ParsePathWithSplitAndReplace :67 ms.
ParsePathWithCustomImplementation :28 ms.
RUN 5:
ParsePathWithSplitAndRegex :128 ms.
ParsePathWithSplitAndReplace :79 ms.
ParsePathWithCustomImplementation :27 ms.
```